src/lightning_app/utilities/app_helpers.py

```diff
@@ -527,3 +528,15 @@ def _should_dispatch_app() -> bool:
         and not bool(int(os.getenv("LIGHTNING_DISPATCHED", "0")))
         and "LIGHTNING_APP_STATE_URL" not in os.environ
     )
+
+
+def _is_headless(app: "LightningApp") -> bool:
+    """Utility which returns True if the given App has no ``Frontend`` objects or URLs exposed through
+    ``configure_layout``."""
+    if app.frontends:
+        return False
+    for component in breadth_first(app.root, types=(lightning_app.LightningFlow,)):
```

Comment:
Unfortunately, an app can be `headless` for some time.
This could reduce the current flexibility of the UI aspects of the framework.
I would prefer to have a property of the root flow to express this instead.

Reply:
I don't think we want to introduce a property here, but it's entirely possible for the `is_headless` flag that we set in the lightning app instance to be mutable. So we would be able to add back support for apps which start off headless and later dynamically start pointing to a URL without needing a flag

Reply:
Spoke with @tchaton - we'll go for updating the `LightningAppInstance` if / when the headless state changes - that will preserve support for dynamic layouts